providers/dns/nicru/nicru.go

```diff
@@ -0,0 +1,329 @@
+package nicru
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/xenolf/lego/acme"
+)
+
+var baseURL = "https://api.nic.ru"
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface that uses
+// DNSMadeEasy's DNS API to manage TXT records for a domain.
+type DNSProvider struct {
+	baseURL     string
+	apiUsername string
+	apiPassword string
+	apiID       string
+	apiSecret   string
+	serviceName string
+	auth        *authResponse
+	client      http.Client
+}
+
+type authResponse struct {
+	ExpiresIn   string `json:"expires_in"`
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+}
+
+type authError struct {
+	Error string `json:"error"`
+}
+
+// Record holds the nic.ru API representation of a DNS Record
+type Record struct {
+	ID       int    `json:"id"`
+	Type     string `json:"type"`
+	Name     string `json:"name"`
+	Value    string `json:"value"`
+	TTL      int    `json:"ttl"`
+	SourceID int    `json:"sourceId"`
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for nic.ru DNS
+func NewDNSProvider() (*DNSProvider, error) {
+	nicruUsername := os.Getenv("NICRU_USERNAME")
+	nicruPassword := os.Getenv("NICRU_PASSWORD")
+	nicruAPIKey := os.Getenv("NICRU_API_KEY")
+	nicruAPISecret := os.Getenv("NICRU_API_SECRET")
+	nicruServiceName := os.Getenv("NICRU_SERVICE_NAME")
+	provider := &DNSProvider{
+		baseURL: baseURL,
+
+		apiUsername: nicruUsername,
+		apiPassword: nicruPassword,
+		apiID:       nicruAPIKey,
+		apiSecret:   nicruAPISecret,
+		serviceName: nicruServiceName,
+
+		client: http.Client{Timeout: 30 * time.Second},
+	}
+
+	err := provider.getAuthRespone()
+	if err != nil {
+		return nil, err
+	}
+	return provider, nil
+}
+
+// getCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for nic.ru
+func (d *DNSProvider) getAuthRespone() error {
+	// TODO: validate all credential are set
+	form := url.Values{}
+	form.Add("grant_type", "password")
+	form.Add("username", d.apiUsername)
+	form.Add("password", d.apiPassword)
+	form.Add("scope", "(GET|PUT|POST|DELETE):/dns-master/.+")
+	form.Add("client_id", d.apiID)
+	form.Add("client_secret", d.apiSecret)
+
+	reqURL := fmt.Sprintf("%s/oauth/token", d.baseURL)
+	req, err := http.NewRequest("POST", reqURL, strings.NewReader(form.Encode()))
+	if err != nil {
+		return err
+	}
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("nicru auth failed, code=%v with unreadable body (%s)", resp.StatusCode, err)
+		}
+
+		var errInfo authError
+		err = json.Unmarshal([]byte(body), &errInfo)
+		if err != nil {
+			return fmt.Errorf("nicru auth failed, code=%v without parsable error: %s, body: %s", resp.StatusCode, err, body)
+		}
```

Comment:
You can simplify this whole section by using `json.NewDecoder(resp.Body).Decode(&errInfo)` like you do below